node/src/node/overseer.cpp

```diff
@@ -459,5 +444,124 @@ overseer_t::on_slave_death(const std::error_code& ec, std::string uuid) {
             pool.erase(it);
         }
     });
-    balancer->on_slave_death(uuid);
+
+    // TODO: Notify watcher about slave death.
+    loop->post(std::bind(&overseer_t::rebalance_slaves, shared_from_this()));
+}
+
+void
+overseer_t::rebalance_events() {
+    using boost::adaptors::filtered;
+
+    const auto concurrency = profile().concurrency;
+
+    pool.apply([&](pool_type& pool) {
+        if (pool.empty()) {
+            return;
+        }
+
+        queue.apply([&](queue_type& queue) {
+            while (!queue.empty()) {
+                // Find an active slave with minimal load, which is not overloaded.
+                const auto filter = std::function<bool(const slave_t&)>([&](const slave_t& slave) -> bool {
+                    return slave.active() && slave.load() < concurrency;
+                });
+
+                const auto range = pool | boost::adaptors::map_values | filtered(filter);
+
+                auto slave = boost::min_element(range, +[](const slave_t& lhs, const slave_t& rhs) -> bool {
+                    return lhs.load() < rhs.load();
+                });
+
+                // No free slaves found.
+                if(slave == boost::end(range)) {
+                    return;
+                }
+
+                auto& payload = queue.front();
+
+                try {
+                    assign(*slave, payload);
+                    // The slave may become invalid and reject the assignment (or reject for any
+                    // other reasons). We pop the channel only on successful assignment to achieve
+                    // strong exception guarantee.
+                    queue.pop_front();
+                } catch (const std::exception& err) {
+                    COCAINE_LOG_DEBUG(log, "slave has rejected assignment: %s", err.what());
+                }
+            }
+        });
+    });
+}
+
+void
+overseer_t::rebalance_slaves() {
+    using boost::adaptors::filtered;
+
+    const auto load = queue->size();
+    const auto profile = this->profile();
+
+    const auto pool_target = this->pool_target.load();
+
+    // Bound current pool target between [1; limit].
+    const auto target = detail::bound(
+        1UL,
```

Comment:
What should I do if there are 0 workers allowed and a new request comes? Drop? Put in the queue? Why did this request come at all?